### ISSUE TITLE
Use XDG standards

### ISSUE THREAD

**Issue by fabiscafe** [probonopd/wayland-x11-compat-protocols#11]
Non-trolling here. Whats wrong with portals? As far as I see it they are the common ground you look to re-implement here without the community behind and under a different name.
So while I do appreciate the take on ease to use, why not just sit on top of a portal?

**Comment by probonopd** [comment]
With X11, you don't need additional software components to get the desired functionality.

Portals essentially come out of the Gtk/Gnome/Flatpak stack with some adoption by KDE. There are many desktop environments which do not have (nor intend to have) the concept of Portals (e.g., because they follow a different philosophy or because they where designed before Portals were conceived).

Hence, at the bare minimum, Wayland should provide the needed functionality natively (and without visible user interfaces) just like X11 _unless Portals are available_.

For example, when an application says "I want to set a global shortcut", then the Wayland compositor could check if there is a Portal for asking the user "Do you really want to set this global shortcut" - or, if there is no such portal installed, _just do it_ (like X11).

In other words, as of today Wayland defaults to being broken (things don't work unless Portals, Pipewire, etc. are there), whereas Wayland should default to being functional (things should "just work") in case Portals, Pipewire, etc. are not there).

**Comment by RushingAlien** [comment]
It's a fair point to make

Wayland is simply part of the freedesktop system and that ecosystem includes the likes of dbus, pipewire and portals
So by removing one of it you are left with an incomplete system, like Linux without a C standard library, shell and coreutils. Wayland is not meant to be standalone and never will.

makes sense to put security sensitive things to portals due to its permission system
makes sense to share compositor framebuffer to pipewire and have pipewire pipe it to app cause pipewire is a multimedia framework that deals with video and audio (also future linux camera stack will use pipewire as well via libcamera)
dbus is just underlying message bus that the likes of portals use

**Comment by fabiscafe** [comment]
@probonopd 
> With X11, you don't need additional software components to get the desired functionality.

Hm I dont think this will bring us anywhere. The desired functionality is a "guarded access". That's something you often cant have on X11 at all. So what you try to compare is at best a subset. But thats what nobody wants and tries to have, because we moved away from X11 for that reason. So let's not go "compared to X11".

> There are many desktop environments which do not have (nor intend to have) the concept of Portals

Currently there are many desktops and windowmanager not sitting on wayland. But is there anything that is in active development and doesn't want to move to wayland?

…

@probonopd From your wording and thinking. Why do you try to build something to circumvent the portals. You clearly describe that you would like to have a generic portal implementation for desktops and compositors to use. Why not build something like this? You know, generic portals that can be hooked onto by desktops without the need to build something themself. Maybe even a simple compositor you can include in your appimages.

Because besides portals as access restriction mechanism, portals also did one thing to the linux desktop we didn't had before: common interfaces. There is no "do I target Qt? Do I target GTK2, 3, 4, maybe GNOME? Maybe something else?" anymore. There is now "I target this portal only" and it will work. It will work on X11, it will work on Wayland.

**Comment by probonopd** [comment]
@RushingAlien 

> Wayland is simply part of the freedesktop system and that ecosystem includes the likes of dbus, pipewire and portals
So by removing one of it you are left with an incomplete system, like Linux without a C standard library, shell and coreutils. 

Not all use cases are about a ecosystem "includes the likes of dbus, pipewire and portals".

> Wayland is not meant to be standalone and never will.

That's what needs to change if Wayland is supposed to become a X11 replacement.

Actually, that's what I call "spaghetti architecture" - when a display server suddenly has a dependency on one particular IPC system (which not everyone wants to use), and one particular sound server (which not everyone wants to use), and one particular way to manage permissions (which not everyone wants to use).

**Comment by probonopd** [comment]
> Hm I dont think this will bring us anywhere. The desired functionality is a "guarded access".

"Desired" by whom? Certainly not by _everyone_, and certainly not in _all use cases_. This whole "guarded access" thing needs to be _optional_.

> That's something you often cant have on X11 at all.

...and I have not asked for it, either.

> So what you try to compare is at best a subset. But thats what nobody wants and tries to have, because we moved away from X11 for that reason. So let's not go "compared to X11".

Actually, the reverse is true (for me): I will only begin to consider switching away from X11 once a replacement exists that lets me do at least what I can do with X11. If that new thing then also can do more, that's a nice to have - but matching X11 core functionality is a must have.

> Currently there are many desktops and windowmanager not sitting on wayland. But is there anything that is in active development and doesn't want to move to wayland?

For example, my young and actively developed desktop environment, helloSystem, will only be able to move to Wayland once there is a known way to have all functionality that it has under X11.

> Why do you try to build something to circumvent the portals.

Because I have no need for them, I have not asked for them, and I don't want the complications (both in terms of usability and code) that come with them.

> You clearly describe that you would like to have a generic portal implementation for desktops and compositors to use. Why not build something like this?

As far as I understand it, Portals expose visible user interfaces to users. That would mean that someone would need to implement every Portal in every GUI toolkit out there, if we want things to look and feel native. And even if someone did invest that time, the result would still be an user experience more difficult than under X11.

> You know, generic portals that can be hooked onto by desktops without the need to build something themself.

So where are all portals for GNUstep, for Enlightenment, etc.? And where are the "no GUI visible" portals?

> Because besides portals as access restriction mechanism, portals also did one thing to the linux desktop we didn't had before: common interfaces. There is no "do I target Qt? Do I target GTK2, 3, 4, maybe GNOME? Maybe something else?" anymore. There is now "I target this portal only" and it will work. It will work on X11, it will work on Wayland.

No it won't, it will break if the Portal is not available on a system. That should be changed.

**Comment by RushingAlien** [comment]
@fabiscafe 
> a generic portal implementation for desktops and compositors to use

There are actually three things close to this
- xdg-desktop-portal-wlr (base portal for all wlroots comp)
- xdg-desktop-portal-xapp (portal by and for xfce, mate and cinnamon)
- xdg-desktop-portal-gtk (used by sway and hyprland in addition to wlr portal to provide file chooser portal, etc. is a dep to xdg-desktop-portal-gnome)

**Comment by fabiscafe** [comment]
@RushingAlien All of them are pretty much project dependent, not generic implementations means if someone decides not to use wlroots as base, it might not work for them to hook up this portal. Correct me if I'm wrong here please. I'm aware that wlroots is like the standard building block for standalone compositors, but if someone is willing to go the route to build a new one, this person needs to also make the whole portal implementation, because there is currently no generic one available.

----------

@probonopd 
> "Desired" by whom? 

Well, every security aware person. Security can't be an optional thing. It needs to be a fundamental design concept. Otherwise, you know, people wont use it and people will try to work this around. People are lazy.

> ...and I have not asked for it, either.

And you're also not moving, but "fighting" the shadow of the past over and over. It's not that I say that everything is dandy. Far from it. But still portals are what we need to have going forward on a desktop.

> I will only begin to consider switching away from X11 once a replacement exists that lets me do at least what I can do with X11.

But you are the one to decide whats possible on you implementation of the wayland protocol, are you? So whats keeps you from implementing the internals as you want and need them to be and the externals in a working-for-portals way?

> Because I have no need for them, I have not asked for them, and I don't want the complications (both in terms of usability and code) that come with them.

I get this. This is however not just you, but lots and lots of people, companies probably some "AI". We all have to work together. That's why we have standards. A common ground to operate and communicate on. The base to move forward and not got stuck in the past.

> Portals expose visible user interfaces to users.

Like a file-chooser, yes. You don't have to take the one from your toolkit. wlroots based ones also can use the GTK stuff, because they don't have an own toolkit. However if you have an toolkit that would be the best option, so every application will use your toolkits native file-chooser instead of a random one.

>  the result would still be an user experience more difficult than under X11.

That's why we need good standards. The process to work them out, yes. That's more difficult, but in the end you will still have one common ground for all Linux apps. Nobody needs to implement desktop specific things and workarounds.

> And where are the "no GUI visible" portals?

Like this https://flatpak.github.io/xdg-desktop-portal/docs/background-app-monitor.html or what do you mean?

> So where are all portals for GNUstep, for Enlightenment, etc.?

You need to ask the projects about this. I'd guess they still are not done with the general wayland port that is sort of a required first step before portal implementation. 

> No it won't, it will break if the Portal is not available on a system.

That's how dependencies work. Make it a dependency and the job is done. 


++ I'll update the title a bit

**Comment by probonopd** [comment]
@fabiscafe 

> But still portals are what we need to have going forward on a desktop

If you think this, then this project is maybe not for you. This project is for those that just want to be able to continue to do everything that we can do in X11, but using Wayland. Hence, this project is about defining the _missing_ Wayland protocols that the upstream wayland-protocols project doesn't want.

> You need to ask the projects about this. I'd guess they still are not done with the general wayland port that is sort of a required first step before portal implementation.

Well, as someone who is working on a desktop environment, I will not even start _thinking_ about porting it to Wayland before Wayland can do the things I need, just like X11, without the need for additional stuff outside of Wayland (e.g., D-Bus, Portals, Pipewire, etc.,). In my desktop environment, I want to use other concepts than what "xdg" uses. I don't want to turn my desktop environment into something that works like Gnome with a different skin.

**Comment by probonopd** [comment]
So no, the display server should not force "xdg" stuff onto us.
